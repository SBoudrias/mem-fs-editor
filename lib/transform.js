const {createTransform} = require('./util');
const {isFilePending} = require('./state');

const createPendingFilesPassthrough = () => createTransform(function (file, _enc, cb) {
  // Don't process deleted file who haven't been commited yet.
  cb(undefined, isFilePending(file) ? file : undefined);
});

const createCommitTransform = memFsEditor => createTransform(function (file, _enc, cb) {
  memFsEditor.commitFileAsync(file).then(() => cb()).catch(error => cb(error));
});

module.exports = {createPendingFilesPassthrough, createCommitTransform};
