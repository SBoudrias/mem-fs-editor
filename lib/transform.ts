import { Transform, TransformCallback } from 'stream';

import { isFilePending } from './state.js';
import type { MemFsEditor, MemFsEditorFile } from './index.js';

export function createTransform<EditorFile extends MemFsEditorFile = MemFsEditorFile>(
  transform: (file: EditorFile, encoding: BufferEncoding, cb: TransformCallback) => void
) {
  return new Transform({
    objectMode: true,
    transform(...args) {
      transform.apply(this, args);
    },
  });
}

export const createPendingFilesPassthrough = () =>
  createTransform((file, _enc, cb) => {
    // Don't process deleted file who haven't been commited yet.
    cb(undefined, isFilePending(file) ? file : undefined);
  });

export const createCommitTransform = <EditorFile extends MemFsEditorFile = MemFsEditorFile>(
  memFsEditor: MemFsEditor<EditorFile>
) =>
  createTransform<EditorFile>((file, _enc, cb) => {
    memFsEditor
      .commitFileAsync(file)
      .then(() => cb())
      .catch((error) => cb(error));
  });
