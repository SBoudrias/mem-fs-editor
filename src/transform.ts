import { Duplex } from 'stream';
import commitFileAsync from './actions/commit-file-async.js';
import type { MemFsEditorFile } from './index.js';

export const createCommitTransform = <EditorFile extends MemFsEditorFile = MemFsEditorFile>() =>
  Duplex.from(async function* (generator: AsyncGenerator<EditorFile>) {
    for await (const file of generator) {
      await commitFileAsync(file);
      yield file;
    }
  });
