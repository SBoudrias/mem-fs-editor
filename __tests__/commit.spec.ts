import { describe, beforeEach, it, expect, afterEach } from 'vitest';
import filesystem from 'fs';
import os from 'os';
import path from 'path';
import { create as createMemFs } from 'mem-fs';
import sinon from 'sinon';
import { MemFsEditor, create } from '../lib/index.js';
import { createTransform } from '../lib/transform.js';
import { getFixture } from './fixtures.js';

const rmSync = filesystem.rmSync || filesystem.rmdirSync;

describe('#commit()', () => {
  const fixtureDir = path.join(os.tmpdir(), '/mem-fs-editor-test-fixture');
  const output = path.join(os.tmpdir(), '/mem-fs-editor-test' + Math.random());
  const NUMBER_FILES = 100;

  let store;
  let fs: MemFsEditor;

  beforeEach(() => {
    store = createMemFs();
    fs = create(store);
    filesystem.mkdirSync(fixtureDir, { recursive: true });

    // Create a 100 files to exercise the stream high water mark
    let i = NUMBER_FILES;
    while (i--) {
      filesystem.writeFileSync(path.join(fixtureDir, 'file-' + i + '.txt'), 'foo');
    }

    fs.copy(fixtureDir + '/**', output);
  });

  afterEach(() => {
    rmSync(fixtureDir, { recursive: true, force: true });
    rmSync(output, { recursive: true, force: true });
  });

  it('should match snapshot', async () => {
    await fs.commit();
    expect(fs.dump(output)).toMatchSnapshot();
  });

  it('call filters and trigger callback on error', async () => {
    let called = 0;

    const filter = createTransform((file, enc, cb) => {
      called++;
      cb(new Error(`error ${called}`));
    });

    await expect(fs.commit([filter])).rejects.toMatch(/error 1/);
  });

  it('call filters and update memory model', async () => {
    let called = 0;

    const filter = createTransform(function (file, enc, cb) {
      called++;
      file.contents = Buffer.from('modified');
      this.push(file);
      cb();
    });

    await fs.commit([filter]);
    expect(called).toBe(100);
    expect(fs.read(path.join(output, 'file-1.txt'))).toBe('modified');
  });

  it('call filters, update memory model and commit selected files', async () => {
    let called = 0;

    const filter = createTransform(function (file, enc, cb) {
      called++;
      file.contents = Buffer.from('modified');
      this.push(file);
      cb();
    });

    const beforeFilter = createTransform(function (file, enc, cb) {
      if (file.path.endsWith('1.txt')) {
        this.push(file);
      }

      cb();
    });

    await fs.commit([filter], store.stream().pipe(beforeFilter));
    expect(called).toBe(10);
    expect(fs.read(path.join(output, 'file-1.txt'))).toBe('modified');
    expect(fs.read(path.join(output, 'file-2.txt'))).not.toBe('modified');
    expect(store.get(path.join(output, 'file-1.txt')).committed).toBeTruthy();
    expect(store.get(path.join(output, 'file-2.txt')).result).toBe(undefined);
  });

  it('write file to disk', async () => {
    await fs.commit();
    expect(filesystem.existsSync(path.join(output, 'file-1.txt'))).toBeTruthy();
    expect(filesystem.existsSync(path.join(output, 'file-1.txt'))).toBeTruthy();
    expect(filesystem.existsSync(path.join(output, 'file-50.txt'))).toBeTruthy();
    expect(filesystem.existsSync(path.join(output, 'file-99.txt'))).toBeTruthy();
  }, 10000);

  it('handle error when write fails', async () => {
    filesystem.writeFileSync(output, 'foo');
    await expect(fs.commit()).rejects.toMatch(/is not a directory/);
  });

  it('delete file from disk', async () => {
    const file = path.join(output, 'delete.txt');
    filesystem.mkdirSync(output, { recursive: true });
    filesystem.writeFileSync(file, 'to delete');

    fs.delete(file);
    await fs.commit();
    expect(filesystem.existsSync(file)).toBeFalsy();
    expect(store.get(file).committed).toBeTruthy();
  });

  it('delete directories from disk', async () => {
    const file = path.join(output, 'nested/delete.txt');
    filesystem.mkdirSync(path.join(output, 'nested'), { recursive: true });
    filesystem.writeFileSync(file, 'to delete');

    fs.delete(path.join(output, 'nested'));
    await fs.commit();
    expect(filesystem.existsSync(file)).toBeFalsy();
  });

  it('reset file status after commiting', async () => {
    await fs.commit();
    expect(fs.store.get(path.join(output, '/file-a.txt')).state).toBeUndefined();
  });

  it('does not commit files who are deleted before being commited', async () => {
    fs.write('to-delete', 'foo');
    fs.delete('to-delete');
    fs.copy(getFixture('file-a.txt'), 'copy-to-delete');
    fs.delete('copy-to-delete');
    fs.store.get('to-delete');

    fs.commitFileAsync = sinon.stub().returns(Promise.resolve());
    await fs.commit([
      createTransform(function (file, enc, cb) {
        expect(file.path).not.toEqual(path.resolve('to-delete'));
        expect(file.path).not.toEqual(path.resolve('copy-to-delete'));

        this.push(file);
        cb();
      }),
    ]);
    expect(fs.commitFileAsync.callCount).toBe(NUMBER_FILES);
  });
});

describe('#copy() and #commit()', () => {
  const output = path.join(os.tmpdir(), '/mem-fs-editor-test');

  let store;
  let fs: MemFsEditor;

  beforeEach(() => {
    store = createMemFs();
    fs = create(store);

    fs.copy(getFixture('**'), output);
  });

  afterEach(() => {
    rmSync(output, { recursive: true, force: true });
  });

  it('should match snapshot', async () => {
    await fs.commit();
    expect(fs.dump(output)).toMatchSnapshot();
  });
});

describe('#copyTpl() and #commit()', () => {
  const output = path.join(os.tmpdir(), '/mem-fs-editor-test');

  let store;
  let fs: MemFsEditor;

  beforeEach(() => {
    store = createMemFs();
    fs = create(store);

    const a = { name: 'foo' } as any;
    const b = { a };
    a.b = b;

    fs.copyTpl(getFixture('**'), output, { name: 'bar' }, { context: { a } });
  });

  afterEach(() => {
    rmSync(output, { recursive: true, force: true });
  });

  it('should match snapshot', async () => {
    await fs.commit();
    expect(fs.dump(output)).toMatchSnapshot();
  });
});
