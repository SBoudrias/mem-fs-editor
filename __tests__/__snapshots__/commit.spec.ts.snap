// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`#commit() > should match snapshot 1`] = `
{
  "file-0.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-1.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-10.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-11.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-12.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-13.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-14.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-15.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-16.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-17.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-18.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-19.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-2.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-20.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-21.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-22.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-23.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-24.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-25.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-26.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-27.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-28.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-29.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-3.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-30.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-31.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-32.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-33.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-34.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-35.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-36.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-37.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-38.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-39.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-4.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-40.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-41.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-42.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-43.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-44.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-45.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-46.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-47.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-48.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-49.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-5.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-50.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-51.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-52.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-53.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-54.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-55.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-56.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-57.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-58.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-59.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-6.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-60.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-61.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-62.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-63.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-64.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-65.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-66.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-67.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-68.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-69.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-7.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-70.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-71.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-72.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-73.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-74.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-75.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-76.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-77.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-78.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-79.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-8.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-80.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-81.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-82.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-83.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-84.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-85.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-86.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-87.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-88.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-89.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-9.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-90.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-91.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-92.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-93.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-94.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-95.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-96.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-97.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-98.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
  "file-99.txt": {
    "contents": "foo",
    "stateCleared": "modified",
  },
}
`;

exports[`#copy() and #commit() > should match snapshot 1`] = `
{
  "[file].txt": {
    "contents": "foo
",
    "stateCleared": "modified",
  },
  "ejs/file-ejs-extension.txt.ejs": {
    "contents": "",
    "stateCleared": "modified",
  },
  "file-(specia!-char$).txt": {
    "contents": "special
",
    "stateCleared": "modified",
  },
  "file-a.txt": {
    "contents": "foo
",
    "stateCleared": "modified",
  },
  "file-binary.bin": {
    "contents": "     ï¿½  \`FFTMf/ï¿½W<%= name %>",
    "stateCleared": "modified",
  },
  "file-binary.jpg": {
    "contents": "     ï¿½  \`FFTMf/ï¿½W<%= name %>",
    "stateCleared": "modified",
  },
  "file-circular.txt": {
    "contents": "<%= this.a.name %> <%= this.a.b.a.name %>
",
    "stateCleared": "modified",
  },
  "file-tpl-custom-delimiter.txt": {
    "contents": "<?= name ?>
",
    "stateCleared": "modified",
  },
  "file-tpl-include.txt": {
    "contents": "<%- include('file-tpl-partial.txt') %>
",
    "stateCleared": "modified",
  },
  "file-tpl-partial.txt": {
    "contents": "partial
",
    "stateCleared": "modified",
  },
  "file-tpl.txt": {
    "contents": "<%= name %>
",
    "stateCleared": "modified",
  },
  "file.json": {
    "contents": "{
  \\"foo\\": \\"bar\\"
}
",
    "stateCleared": "modified",
  },
  "nested/file.txt": {
    "contents": "nested
",
    "stateCleared": "modified",
  },
}
`;

exports[`#copyTpl() and #commit() > should match snapshot 1`] = `
{
  "[file].txt": {
    "contents": "foo
",
    "stateCleared": "modified",
  },
  "ejs/file-ejs-extension.txt": {
    "contents": "",
    "stateCleared": "modified",
  },
  "file-(specia!-char$).txt": {
    "contents": "special
",
    "stateCleared": "modified",
  },
  "file-a.txt": {
    "contents": "foo
",
    "stateCleared": "modified",
  },
  "file-binary.bin": {
    "contents": "     ï¿½  \`FFTMf/ï¿½W<%= name %>",
    "stateCleared": "modified",
  },
  "file-binary.jpg": {
    "contents": "     ï¿½  \`FFTMf/ï¿½W<%= name %>",
    "stateCleared": "modified",
  },
  "file-circular.txt": {
    "contents": "foo foo
",
    "stateCleared": "modified",
  },
  "file-tpl-custom-delimiter.txt": {
    "contents": "<?= name ?>
",
    "stateCleared": "modified",
  },
  "file-tpl-include.txt": {
    "contents": "partial

",
    "stateCleared": "modified",
  },
  "file-tpl-partial.txt": {
    "contents": "partial
",
    "stateCleared": "modified",
  },
  "file-tpl.txt": {
    "contents": "bar
",
    "stateCleared": "modified",
  },
  "file.json": {
    "contents": "{
  \\"foo\\": \\"bar\\"
}
",
    "stateCleared": "modified",
  },
  "nested/file.txt": {
    "contents": "nested
",
    "stateCleared": "modified",
  },
}
`;
